File: src/python/taxpulse_python/tax_liability.py
from collections.abc import Awaitable
from typing import Protocol

from pydantic import ConfigDict, ValidationError

from taxpulse_python.tax_liability_model import (
    StrictTaxPulseModel,
    TaxCalcFilingStatus,
    TaxCalcRequest,
    TaxPlanCycleAggregationPayload,
)


class TaxLiabilitySourceError(Exception):
    pass


class TaxLiabilityValidationError(Exception):
    def __init__(self, error: ValidationError) -> None:
        super().__init__("Invalid Tax Plan Cycle payload for real-time tax-liability calculation.")
        self.error = error


class StrictFrozenTaxPulseResult(StrictTaxPulseModel):
    model_config = ConfigDict(extra="forbid", strict=True, frozen=True)


class TaxCalcResult(StrictFrozenTaxPulseResult):
    filing_status: TaxCalcFilingStatus
    state: str
    taxable_income: float
    estimated_tax_liability: float
    is_placeholder: bool


class TaxLiabilityScenarioResult(StrictFrozenTaxPulseResult):
    scenario_id: str
    label: str
    projected_tax_liability: float
    projected_net_income_after_deductions: float

    def as_dict(self) -> dict[str, str | float]:
        return {
            "scenario_id": self.scenario_id,
            "label": self.label,
            "projected_tax_liability": self.projected_tax_liability,
            "projected_net_income_after_deductions": self.projected_net_income_after_deductions,
        }


class TaxLiabilityModelingResult(StrictFrozenTaxPulseResult):
    payload: TaxPlanCycleAggregationPayload
    total_holding_market_value: float
    scenario_results: tuple[TaxLiabilityScenarioResult, ...]


class TaxPlanCycleDataSource(Protocol):
    def load_payload(self) -> Awaitable[object]: ...


async def calculate_tax_result(request: TaxCalcRequest) -> TaxCalcResult:
    taxable_income = max(request.income - request.deductions, 0.0)

    return TaxCalcResult(
        filing_status=request.filing_status,
        state=request.state,
        taxable_income=taxable_income,
        estimated_tax_liability=taxable_income * 0.20,
        is_placeholder=True,
    )


def _calculate_scenario_results(
    payload: TaxPlanCycleAggregationPayload,
) -> tuple[TaxLiabilityScenarioResult, ...]:
    base_income = sum(item.amount for item in payload.income_events)
    base_deductions = sum(item.amount for item in payload.deductions)

    results: list[TaxLiabilityScenarioResult] = []
    for scenario in payload.scenarios:
        projected_income = base_income + scenario.projected_additional_income
        projected_deductions = base_deductions + scenario.projected_additional_deductions
        projected_net_income_after_deductions = projected_income - projected_deductions
        taxable_amount = max(projected_net_income_after_deductions, 0)
        results.append(
            TaxLiabilityScenarioResult(
                scenario_id=scenario.scenario_id,
                label=scenario.label,
                projected_tax_liability=taxable_amount * scenario.effective_tax_rate,
                projected_net_income_after_deductions=projected_net_income_after_deductions,
            )
        )

    return tuple(results)


async def model_tax_liability(source: TaxPlanCycleDataSource) -> TaxLiabilityModelingResult:
    try:
        raw_payload = await source.load_payload()
    except Exception as error:
        raise TaxLiabilitySourceError(
            "Unable to load Tax Plan Cycle payload for real-time tax-liability calculation."
        ) from error

    try:
        payload = TaxPlanCycleAggregationPayload.model_validate(raw_payload)
    except ValidationError as error:
        raise TaxLiabilityValidationError(error) from error

    return TaxLiabilityModelingResult(
        payload=payload,
        total_holding_market_value=sum(holding.market_value for holding in payload.holdings),
        scenario_results=_calculate_scenario_results(payload),
    )

File: src/typescript/tax-liability.ts
import { ZodError, type ZodType } from "zod";
import {
  DeductionSchema,
  HoldingSchema,
  IncomeEventSchema,
  TaxLiabilityScenarioInputSchema,
  type Deduction,
  type Holding,
  type IncomeEvent,
  type TaxLiabilityScenarioInput,
  type TaxPlanCycleAggregationPayload
} from "./tax-liability.schema.js";

export class TaxLiabilityTimeoutError extends Error {
  public constructor(timeoutMs: number) {
    super(`Timed out after ${timeoutMs}ms while loading Tax Plan Cycle data.`);
    this.name = "TaxLiabilityTimeoutError";
  }
}

export class TaxLiabilityValidationError extends Error {
  public readonly issues: ZodError["issues"];

  public constructor(
    public readonly panel: TaxPlanCycleDataPanel,
    error: ZodError
  ) {
    super(`Invalid ${panel} payload for Tax Plan Cycle modeling.`);
    this.name = "TaxLiabilityValidationError";
    this.issues = error.issues;
  }
}

export class TaxLiabilitySourceError extends Error {
  public constructor(
    public readonly panel: TaxPlanCycleDataPanel,
    cause: unknown
  ) {
    super(`Unable to load ${panel} payload for Tax Plan Cycle modeling.`, { cause });
    this.name = "TaxLiabilitySourceError";
  }
}

export type TaxPlanCycleDataPanel = "incomeEvents" | "deductions" | "holdings" | "scenarios";

export type PanelOutcome<Value> =
  | {
      readonly status: "fulfilled";
      readonly value: Value;
    }
  | {
      readonly status: "rejected";
      readonly reason:
        | TaxLiabilitySourceError
        | TaxLiabilityTimeoutError
        | TaxLiabilityValidationError;
    };

export interface TaxPlanCycleInputCollection {
  readonly incomeEvents: PanelOutcome<IncomeEvent[]>;
  readonly deductions: PanelOutcome<Deduction[]>;
  readonly holdings: PanelOutcome<Holding[]>;
  readonly scenarios: PanelOutcome<TaxLiabilityScenarioInput[]>;
}

export interface TaxLiabilityScenarioResult {
  readonly scenarioId: string;
  readonly label: string;
  readonly projectedTaxLiability: number;
  readonly projectedNetIncomeAfterDeductions: number;
}

export interface TaxPlanCycleModelingResult {
  readonly payload: TaxPlanCycleAggregationPayload;
  readonly totalHoldingMarketValue: number;
  readonly scenarioResults: TaxLiabilityScenarioResult[];
}

export interface TaxPlanCycleDataSource {
  readonly loadIncomeEvents: () => Promise<unknown>;
  readonly loadDeductions: () => Promise<unknown>;
  readonly loadHoldings: () => Promise<unknown>;
  readonly loadScenarios: () => Promise<unknown>;
}

export interface MockTaxPlanCycleDataSourceOptions {
  readonly payload: TaxPlanCycleAggregationPayload;
  readonly latencyMsByPanel?: Partial<Record<TaxPlanCycleDataPanel, number>>;
  readonly rejectedPanels?: Partial<Record<TaxPlanCycleDataPanel, unknown>>;
}

export interface TaxPlanCycleModelingClient {
  readonly collectInputs: (timeoutMs: number) => Promise<TaxPlanCycleInputCollection>;
  readonly modelScenarios: (timeoutMs: number) => Promise<TaxPlanCycleModelingResult>;
}

const sleep = async (durationMs: number): Promise<void> => {
  await new Promise<void>((resolve) => {
    setTimeout(resolve, durationMs);
  });
};

const withTimeout = async <Result>(work: Promise<Result>, timeoutMs: number): Promise<Result> => {
  let timeoutId: ReturnType<typeof setTimeout> | undefined;

  const timeout = new Promise<never>((_, reject) => {
    timeoutId = setTimeout(() => {
      reject(new TaxLiabilityTimeoutError(timeoutMs));
    }, timeoutMs);
  });

  try {
    return await Promise.race([work, timeout]);
  } finally {
    if (timeoutId !== undefined) {
      clearTimeout(timeoutId);
    }
  }
};

const normalizePanelError = (
  panel: TaxPlanCycleDataPanel,
  error: unknown
): TaxLiabilitySourceError | TaxLiabilityTimeoutError | TaxLiabilityValidationError => {
  if (error instanceof TaxLiabilityTimeoutError) {
    return error;
  }

  if (error instanceof TaxLiabilityValidationError) {
    return error;
  }

  if (error instanceof ZodError) {
    return new TaxLiabilityValidationError(panel, error);
  }

  return new TaxLiabilitySourceError(panel, error);
};

const loadPanel = async <Value>(
  panel: TaxPlanCycleDataPanel,
  loader: () => Promise<unknown>,
  schema: ZodType<Value>,
  timeoutMs: number
): Promise<Value> => {
  const rawPayload = await withTimeout(loader(), timeoutMs);
  return schema.parse(rawPayload);
};

const toPanelOutcome = <Value>(
  panel: TaxPlanCycleDataPanel,
  result: PromiseSettledResult<Value>
): PanelOutcome<Value> => {
  if (result.status === "fulfilled") {
    return {
      status: "fulfilled",
      value: result.value
    };
  }

  return {
    status: "rejected",
    reason: normalizePanelError(panel, result.reason)
  };
};

const requirePanel = <Value>(outcome: PanelOutcome<Value>): Value => {
  if (outcome.status === "fulfilled") {
    return outcome.value;
  }

  throw outcome.reason;
};

const sumByAmount = <Item extends { readonly amount: number }>(items: Item[]): number =>
  items.reduce((total, item) => total + item.amount, 0);

const calculateScenarioResults = (
  incomeEvents: IncomeEvent[],
  deductions: Deduction[],
  scenarios: TaxLiabilityScenarioInput[]
): TaxLiabilityScenarioResult[] => {
  const baseIncome = sumByAmount(incomeEvents);
  const baseDeductions = sumByAmount(deductions);

  return scenarios.map((scenario) => {
    const projectedIncome = baseIncome + scenario.projectedAdditionalIncome;
    const projectedDeductions = baseDeductions + scenario.projectedAdditionalDeductions;
    const projectedNetIncomeAfterDeductions = projectedIncome - projectedDeductions;
    const taxableAmount = Math.max(projectedNetIncomeAfterDeductions, 0);

    return {
      scenarioId: scenario.scenarioId,
      label: scenario.label,
      projectedTaxLiability: taxableAmount * scenario.effectiveTaxRate,
      projectedNetIncomeAfterDeductions
    };
  });
};

const sumHoldingMarketValue = (holdings: Holding[]): number =>
  holdings.reduce((total, holding) => total + holding.marketValue, 0);

export const createMockTaxPlanCycleDataSource = ({
  payload,
  latencyMsByPanel = {},
  rejectedPanels = {}
}: MockTaxPlanCycleDataSourceOptions): TaxPlanCycleDataSource => {
  const loadPanelPayload = async (panel: TaxPlanCycleDataPanel): Promise<unknown> => {
    await sleep(latencyMsByPanel[panel] ?? 0);

    if (panel in rejectedPanels) {
      throw rejectedPanels[panel];
    }

    return payload[panel];
  };

  return {
    loadIncomeEvents: async (): Promise<unknown> => loadPanelPayload("incomeEvents"),
    loadDeductions: async (): Promise<unknown> => loadPanelPayload("deductions"),
    loadHoldings: async (): Promise<unknown> => loadPanelPayload("holdings"),
    loadScenarios: async (): Promise<unknown> => loadPanelPayload("scenarios")
  };
};

export const createTaxPlanCycleModelingClient = (
  source: TaxPlanCycleDataSource,
  taxPlanCycle: TaxPlanCycleAggregationPayload["taxPlanCycle"]
): TaxPlanCycleModelingClient => {
  const collectInputs = async (timeoutMs: number): Promise<TaxPlanCycleInputCollection> => {
    const [incomeEvents, deductions, holdings, scenarios] = await Promise.allSettled([
      loadPanel("incomeEvents", source.loadIncomeEvents, IncomeEventSchema.array(), timeoutMs),
      loadPanel("deductions", source.loadDeductions, DeductionSchema.array(), timeoutMs),
      loadPanel("holdings", source.loadHoldings, HoldingSchema.array(), timeoutMs),
      loadPanel(
        "scenarios",
        source.loadScenarios,
        TaxLiabilityScenarioInputSchema.array().min(2).max(5),
        timeoutMs
      )
    ]);

    return {
      incomeEvents: toPanelOutcome("incomeEvents", incomeEvents),
      deductions: toPanelOutcome("deductions", deductions),
      holdings: toPanelOutcome("holdings", holdings),
      scenarios: toPanelOutcome("scenarios", scenarios)
    };
  };

  return {
    collectInputs,
    modelScenarios: async (timeoutMs: number): Promise<TaxPlanCycleModelingResult> => {
      const inputs = await collectInputs(timeoutMs);
      const incomeEvents = requirePanel(inputs.incomeEvents);
      const deductions = requirePanel(inputs.deductions);
      const holdings = requirePanel(inputs.holdings);
      const scenarios = requirePanel(inputs.scenarios);

      const payload = {
        taxPlanCycle,
        incomeEvents,
        deductions,
        holdings,
        scenarios
      } satisfies TaxPlanCycleAggregationPayload;

      return {
        payload,
        totalHoldingMarketValue: sumHoldingMarketValue(holdings),
        scenarioResults: calculateScenarioResults(incomeEvents, deductions, scenarios)
      };
    }
  };
};
